Add 1D data visualization tool

isha1810 opened the pull request:
## Proposed changes
The python file contains code to visualize 1D data using matplotlib and saves the animation to mpeg file.

### Types of changes:

- [ ] Bugfix
- [x] New feature

### Component:

- [x] Code
- [ ] Documentation
- [ ] Build system
- [ ] Continuous integration

### Code review checklist

- [ ] The PR passes all checks, including unit tests and `clang-tidy`.
  For instructions on how to perform the CI checks locally refer to the [Dev
  guide on the Travis CI](https://spectre-code.org/travis_guide.html).
- [ ] The code is documented and the documentation renders correctly. Run
  `make doc` to generate the documentation locally into `BUILD_DIR/docs/html`.
  Then open `index.html`.
- [ ] The code follows the stylistic and code quality guidelines listed in the
  [code review guide](https://spectre-code.org/code_review_guide.html).

### Further comments

<!--
If this is a relatively large or complex change, kick off the discussion by
explaining why you chose the solution you did and what alternatives you
considered, etc...
-->

isha1810 committed:
Added 1D data visualisation using python

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+#!/usr/bin/env python
```
This line has to be first.  You should also make the file executable with `chomd +x Render1D.py` or similar.

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+    :param arr: the array over which to find the max and min
+    This doesn't really work if there are nan valued data pts.
+    '''
+    # The last statment of the docstring may or may not be true
```
You will need to figure this out.

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+                        help='set to the name of the output file you want \
+                        written. For animations this saves an mp4 file and \
+                        for stills a pdf. (optional)')
+    parser.add_argument('--fps', type=str, required=False,
```
Shouldn't this be a float?  Check the other arguments to make sure they make sense too.

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+    ax.set_xlabel('x')
+    ax.set_ylabel(args['var'])
+    title = ax.set_title("")
+    my_interval = 200
```
Set the defaults when you define the options.

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+    parser = argparse.ArgumentParser(description='Render 1-dimensional data',
+                                     formatter_class=
+                                     argparse.ArgumentDefaultsHelpFormatter)
+    parser.add_argument('--file-prefix', type=str, required=True,
```
Instead of taking a prefix and globbing, could you just take the files as positional arguments?  That would be more flexible.

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+                        instead of an animation (optional)')
+    parser.add_argument('--save', type=str, required=False,
+                        help='set to the name of the output file you want \
+                        written. For animations this saves an mp4 file and \
```
It says mpg in the code.

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+    parser.add_argument('--time', type=int,required=False,
+                        help='if given renders the integer observation step \
+                        instead of an animation (optional)')
+    parser.add_argument('--save', type=str, required=False,
```
The places this is used add the appropriate extension.  This behavior needs to be documented or removed.  (My preference would be to just write to the name given and depend on the user to give the correct extension.  This avoids files like foo.pdf.pdf that I tend to get from SpEC.)

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+    volfiles =  [h5file['element_data.vol'] for h5file in h5files]
+    # Get a list of times from the first vol file
+    ids_times = [(obs_id,volfiles[0][obs_id].attrs['observation_value']) for
+                 obs_id in volfiles[0].keys()]
```
You need to handle the case where the user makes a typo and there are no matching h5 files.

nilsdeppe replied:
I think this is now handled by `get_h5_files` and `argparse` together. Is that correct, @isha1810 ?

wthrowe reviewed (changes requested):
I don't know python very well, so I can't really critique the code, but I have some comments on the user interface.

nilsvu commented:
Is this script being added to the main repository for a particular purpose? It looks like a better place for it right now would be https://github.com/sxs-collaboration/spectre_vizualization, which collects scripts like this for use by other people. To make its way into the main spectre repository, I think it needs the following:
- It should be placed in the `spectre` Python module, e.g. in a new `spectre.visualization` module
- It needs unit tests to make sure it keeps working

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+    h5files = get_h5_files(file_prefix)
+    volfile = h5files[0]["element_data.vol"]
+    obs_id_0  = volfile.keys()[0]
+    grid_names = volfile[Observation_Id_0].keys()
```
`Observation_Id_0` is never defined.  Presumably it is supposed to be the same variable defined on the previous line.

wthrowe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,255 @@
+    grid = volfile[Observation_Id_0][grid_names[0]]
+    variables = grid.keys()
+    variables.remove("connectivity")
+    variables.remove("x-coord")
```
This did not exist in the file I had.

wthrowe reviewed (commented):
I tried the script out, and I found that some of the functionality did not work.  In addition to the line comments, I got errors when running under python 3.

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+mpl.use('Agg')
+import matplotlib.pyplot as plt
+import matplotlib.animation as animation
+try:
```
Why do you need to import tkinter explicitly? I've never had to do this, but maybe my code has been wrong all this time :man_shrugging:

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+
+    group= parser.add_mutually_exclusive_group(required=True)
+    group.add_argument('--file-prefix', type=str,
+                        help='common prefix of all h5 files being used'
```
Can you double check that you don't need a space at the end of this line in the quotes? I'm worried this will get rendered as: `files being usedin the case`. Here It looks like you actually want `used. ` (space if needed), then start a new sentence

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+                        'without the .h5 extension')
+    group.add_argument('--filename-list', type=str, nargs = '+',
+                       help='list of all files if they do not'
+                       'have same file_prefix name. Include \'.h5\' extension')
```
`file_prefix name` -> `file prefix` I think is clearer.

`Include` -> `You must include the extension, e.g. \'.h5\'`

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+                       'have same file_prefix name. Include \'.h5\' extension')
+
+    parser.add_argument('--var', type=str, required=True,
+                        help='name of variable to render. eg \'psi'
```
eg -> E.g.

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+                        help='name of variable to render. eg \'psi'
+                        'psi-Error\' or \'psi-Analytic\'')
+    parser.add_argument('--time', type=int,required=False,
+                        help='if given renders the integer observation step'
```
given -> specified

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+                        'psi-Error\' or \'psi-Analytic\'')
+    parser.add_argument('--time', type=int,required=False,
+                        help='if given renders the integer observation step'
+                        'instead of an animation (optional)')
```
I think arg parse will automatically mark the option as optional. Am I misremembering?

isha1810 replied:
No, it doesn't print it, only prints default value specified.

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+                        help='if given renders the integer observation step'
+                        'instead of an animation (optional)')
+    parser.add_argument('--save', type=str, required=False,
+                        help='set to the name of the output file you want'
```
A couple thoughts: maybe we should use `-o` and `--output` instead of `--save` since `-o` and `--output` are certainly more common.

Maybe: `Set the name of the output file. If not specified renders to screen. ....`

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+                        help='set to the name of the output file you want'
+                        'written. For animations this saves an mp4 file and'
+                        'for stills a pdf. Name of file should not include'
+                        'file extension (optional)')
```
same question about `(optional)`, and for the args below

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+                        help='Set the number of frames per second when writing'
+                        'an animation to disk.(optional)')
+    parser.add_argument('--interval', type=int, required=False, default = 200,
+                        help='number of milliseconds (optional)')
```
It's good you gave the units since that's often a source of confusion :) Could you add a brief description of what this is controlling?

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+    parser.add_argument('--interval', type=int, required=False, default = 200,
+                        help='number of milliseconds (optional)')
+    parser.add_argument('--list-vars', type=bool, required=False,
+                        help='if true then write list of vars in h5 file'
```
maybe `write` -> `print to screen`

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+    if parser.parse_args().file_prefix is not None and \
+       parser.parse_args().file_prefix.endswith('.h5'):
+        print("Retry without .h5 extension on 'file-prefix'")
+        exit()
```
you should probably exit with a non-zero error code.

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+            h5_files = [h5py.File(file_name, 'r')
+                        for file_name in filename_list]
+        except FileNotFoundError:
+            exit("File was not found in current directory")
```
Can you somehow print which file was not found? If I'm requesting 10 files then this error message would leave me a bit frustrated :) Also, can you print out what the current directory is?

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+                        for file_name in filename_list]
+        except FileNotFoundError:
+            exit("File was not found in current directory")
+    return(h5_files)
```
I think you can just do `return h5_files`, or is this meant to create a one-element tuple?

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+def get_data(args):
+    '''
+    Get the data to be plotted
+    #:return: the array of data, coords, and time
```
Why is this octothorpe here? (`#`)

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+            h5_files = [h5py.File(file_name, 'r')
+                        for file_name in h5_file_names]
+        except FileNotFoundError:
+            exit("File was not found in current directory")
```
Maybe `Could not find files with prefix PREFIX in directory DIR`?

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+        local_data = []
+        local_coords = []
+        for volfile in volfiles:
+            Obs_Id_File  = volfile[id_and_time[0]]
```
`obs_id_file` (no capital letters)

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+if __name__ == "__main__":
+    try:
+        main(parse_cmd_line())
+    except KeyboardInterrupt:
```
What does this do? If I use `Control-C` I would expect the program to be killed. Does that still happen?

isha1810 replied:
Yeah, Control-C kills the program.

nilsdeppe replied:
What I'm still not understanding is why the try-except is needed or what it enables that otherwise wouldn't work. Could you please explain to me what happens differently compared to not having it?

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+    '''
+    plt.xlabel("x")
+    plt.ylabel(args['var'])
+    plt.title("t = %.5f" % time[int(args['time'])])
```
I'm confused. Where is the `time` variable defined? It's not one of the arguments

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,253 @@
+              (args['save'], myfps))
+        anim.save(args['save'] + ".mp4" , writer='ffmpeg')
+    else:
+        def propagate_exceptions(self, exc, val, tb):
```
A comment on what this exception code is doing would be useful :)

nilsdeppe replied:
ping on adding the comment

nilsdeppe reviewed (changes requested):
Looks good! I'll have to test it out and see if I can break it somehow. Is there a way to add tests for some of the functions?

nilsdeppe commented:
I'm happy with the changes. Could you rebase on develop and squash the fixups into the main commit? I'm happy to do this together at our meeting on Wednesday if that would be helpful :)

nilsdeppe commented on tools/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+#!/usr/bin/env python
```
This looks good. Last couple remaining things:
- Can you move this to `src/Visualization/Python` and add to the CMake file there:
  ```cmake
  spectre_python_add_executable(
    Render1D
    Visualization/Render1D.py
    )
  ```
  as well as add it to the list of `PYTHON_FILES`?
- Once you've moved it, can you make this file executable by running `chmod +x Render1D.py`?
- Your commit message is currently:
  ```
  Added 1D data visualisation using python

  fixup
  ```
  Can you remove the "fixup"? This can be done with an interactive rebase where instead of doing "s" or "f" you do "r" for "reword". We can do this in person on Friday too :)

nilsdeppe replied:
You're still adding this file into `tools` in addition to `src/Visualization/python`

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    plt.title("t = %.5f" % int(args['time']))
+    plt.plot(coords[int(args['time'])], data[int(args['time'])], 'o')
+    if args['save']:
+        print("Writing still to file %s.pdf" % args['save'])
```
Please use Python‘s `logging` module for these kind of status messages. Essentially you just replace `print` with `logger.info` and retrieve an appropriate logger at the beginning of the file or at the beginning of the function, e.g. as `logger = logging.getLogger(__name__)`. You can find examples how to do this in our other Python scripts, e.g. CleanOutput.py or GenerateXdmf.py

nilsvu reviewed (changes requested):
Not a full review, just read through the PR and noticed this. I‘ll wait until you have determined this is not _in progress_ anymore before reviewing :)

nilsdeppe reviewed (dismissed):
I'm happy with this

nilsdeppe commented:
@wthrowe would you mind taking a look at this again?

nilsdeppe commented:
@wthrowe and @nilsleiffischer could you please review this or step down as reviewers if you don't want to review? Thanks!

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+                        '(optional)')
+    parser.add_argument(
+        '--fps',
+        type=int,
```
This option doesn't work.  I get the same frame rate no matter what I pass.

Also, shouldn't this be a float?

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+                        help='Name of variable to render. E.g. \'psi '
+                        'psi-Error\' or \'psi-Analytic\'')
+    parser.add_argument(
+        '--time',
```
This option does not work.  If I pass '0' it seems to be completely ignored and saves an animation anyway.  If I pass a higher number I get a backtrace:
```
Traceback (most recent call last):
  File "/home/wthrowe/spectre/src/Visualization/Python/Render1D.py", line 277, in <module>
    main(parse_cmd_line())
  File "/home/wthrowe/spectre/src/Visualization/Python/Render1D.py", line 270, in main
    render_single_time(args, coords, data)
  File "/home/wthrowe/spectre/src/Visualization/Python/Render1D.py", line 194, in render_single_time
    if args['save']:
KeyError: 'save'
```

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+        required=False,
+        default=200,
+        help='Delay between frames in  milliseconds (optional)')
+    parser.add_argument('--list-vars',
```
This errors unless I also pass `--var`.  This is supposed to tell me the valid values for `--var`, so I should be able to call it without knowing what to pass to `--var`.

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+                        type=str,
+                        required=True,
+                        help='Name of variable to render. E.g. \'psi '
+                        'psi-Error\' or \'psi-Analytic\'')
```
We don't name our variables this way.  We use things like `Error(U)`.

wthrowe replied:
The documentation claims I can pass multiple values (`'psi psi-Error'`), but that does not seem to work.

wthrowe replied:
Passing an invalid value should give a nicer error than
```
Traceback (most recent call last):
  File "/home/wthrowe/spectre/src/Visualization/Python/Render1D.py", line 277, in <module>
    main(parse_cmd_line())
  File "/home/wthrowe/spectre/src/Visualization/Python/Render1D.py", line 268, in main
    time, coords, data = get_data(args)
  File "/home/wthrowe/spectre/src/Visualization/Python/Render1D.py", line 169, in get_data
    local_data += list(obs_id_file[var_name])
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "/usr/lib64/python3.6/site-packages/h5py/_hl/group.py", line 169, in __getitem__
    oid = h5o.open(self.id, self._e(name), lapl=self._lapl)
  File "h5py/_objects.pyx", line 54, in h5py._objects.with_phil.wrapper
  File "h5py/_objects.pyx", line 55, in h5py._objects.with_phil.wrapper
  File "h5py/h5o.pyx", line 190, in h5py.h5o.open
KeyError: "Unable to open object (Object 'u error(u)' doesn't exist)"
```

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+                                     formatter_class=
+                                     argparse.ArgumentDefaultsHelpFormatter)
+    group = parser.add_mutually_exclusive_group(required=True)
+    group.add_argument('--file-prefix',
```
If I give a name that does not match anything it prints "Could not find files with prefix BurgersVolume0z in path".  What does "in path" mean?

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+                       help='Common prefix of all h5 files being used '
+                       'in the case of a single h5 file, it is the file name '
+                       'without the .h5 extension')
+    group.add_argument('--filename-list',
```
Similar to `--file-prefix`, on bad input this prints "Could not find file BurgersVolume0.h5zzz in directory" without telling me what directory.

wthrowe reviewed (changes requested):
I haven't reviewed the code, only observed the behavior when I called the script.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+
+# Distributed under the MIT License.
+# See LICENSE.txt for details.
+# Python script to render 1D data written using SpECTRE in format v2
```
Where is this `v2` coming from? Have we actually incremented a version number somewhere?

nilsvu replied:
ping

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+import os
+import numpy as np
+import matplotlib as mpl
+mpl.use('Agg')
```
what do you need this for? It makes your `plt.show()` calls disfunctional

nilsdeppe replied:
I think this is needed on systems without a window system, e.g. supercomputers. Do you know of a good way to handle both cases?

nilsvu replied:
on minerva this script runs without error:

```py
import matplotlib.pyplot as plt
plt.plot(range(10))
plt.savefig('test.pdf')
```

did you have issues with this on wheeler?

nilsdeppe replied:
```
  File "./test.py", line 2, in <module>
    plt.plot(range(10))
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/pyplot.py", line 3306, in plot
    ax = gca()
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/pyplot.py", line 950, in gca
    return gcf().gca(**kwargs)
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/pyplot.py", line 586, in gcf
    return figure()
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/pyplot.py", line 535, in figure
    **kwargs)
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt5agg.py", line 44, in new_figure_manager
    return new_figure_manager_given_figure(num, thisFig)
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt5agg.py", line 51, in new_figure_manager_given_figure
    canvas = FigureCanvasQTAgg(figure)
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt5agg.py", line 242, in __init__
    super(FigureCanvasQTAgg, self).__init__(figure=figure)
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt5agg.py", line 66, in __init__
    super(FigureCanvasQTAggBase, self).__init__(figure=figure)
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt5.py", line 236, in __init__
    _create_qApp()
  File "/usr/local/python/anaconda2-4.4.0/lib/python2.7/site-packages/matplotlib/backends/backend_qt5.py", line 144, in _create_qApp
    raise RuntimeError('Invalid DISPLAY variable')
RuntimeError: Invalid DISPLAY variable
```
and this is something I see on every supercomputer I've ever used... So I'm surprised you don't see it on Minerva

nilsvu replied:
- It's up to the machine, not the script, which backend should be used. There are several ways to set the backend, the easiest is the `MPLBACKEND` env variable:
  ```
  MPLBACKEND="Agg" path/to/Render1D.py
  ```
  See [the docs](https://matplotlib.org/faq/usage_faq.html#what-is-a-backend) for details.
- Newer versions of Matplotlib set a reasonable backend by default, so if you're running this in Py3 on Wheeler you should be fine without configuring anything (this is what I did on Minerva).

nilsdeppe replied:
Unfortunately I don't think we can rely on the machine to be configured correctly... We need to be able to handle this at the script level

nilsvu replied:
What's wrong with the env variable?

nilsdeppe replied:
Nothing, it's just we cannot rely on the machines to set them, and I don't think it's very user friendly to have people invoke it that way...

nilsvu replied:
They only have to invoke it with the env var if they're not using Py3, which they should be :D  Py2 has reached EOL! Is Py3 unavailable on Wheeler?

The issue is that I don't see a good way to set the backend in the script while still supporting both showing a window _and_ plotting to a file on any machine. The issue is that (if I understand correctly) the backend has to be set before importing `matplotlib.pyplot`, and unless that's being done _within_ the plotting function we can't set the backend in `main` depending on whether the user provided a save-to filename.

nilsdeppe replied:
Python 2 will be used in the scientific community for the next several years, that's just the reality of things. We cannot require python 3 yet.

Since your the python expert, it would be nice if you figured out the solution here... We really cannot rely on anything but our own script to handle these cases... One option is just to not plot directly and always print to file.

nilsvu replied:
Here are the options (that I can think of now). I let @wthrowe and you decide which one to follow:
1. Move the `import matplotlib.pyplot as plt` into the two `render_` functions and set a specific backend in `main`:

    ```py
    if args.output is None:
        mpl.use("TkAgg")
    else:
        mpl.use("Agg")
    ```
    Keep this in mind (from [here](https://matplotlib.org/faq/usage_faq.html#what-is-a-backend)):
    > Using `use()` will require changes in your code if users want to use a different backend. Therefore, you should avoid explicitly calling `use()` unless absolutely necessary.

2. Delete lines 13-14. Running the script with Py3 works by default. Running with Py2 may complain about the backend, in which case the user should run the script as `MPLBACKEND="Agg" Render1D.py`.

nilsdeppe replied:
In the first, can we not also check that the `MPILBACKEND` variable is unset and that we are using python2? Doesn't that cover the downsides you're worried about?

nilsvu replied:
The user may have a backend configured in their `~/.config/matplotlib` or similar config file, so just checking the env var is not enough. I'm not aware of a way to check if the `matplotlib.get_backend()` is the default or if it's configured by the user.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    '''
+    Find max and min over a range of number arrays
+    :param arr: the array over which to find the max and min
+    This doesn't really work if there are nan valued data pts.
```
What do you mean with it "doesn't really work"? It'll just ignore the NaNs if I understand correctly

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    :return: dictionary of the command-line args, dashes are underscores
+    '''
+
+    parser = argparse.ArgumentParser(\
```
Delete the `\` ([E502](https://www.flake8rules.com/rules/E502.html))

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+                        'written. For animations this saves an mp4 file and '
+                        'for stills a pdf. Name of file should not include '
+                        'file extension. If not specified, renders to screen. '
+                        '(optional)')
```
You can omit the `(optional)` in these texts. It's listed as "optional" automatically.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+        type=int,
+        required=False,
+        default=200,
+        help='Delay between frames in  milliseconds (optional)')
```
Is `interval` an alternative to `fps`? If so, you can also make it a "group" as you have done for the file names. But perhaps `fps` is sufficient and you can just remove the `interval` alternative

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+                        help='If true then print to screen vars in h5 file '
+                        'and exit (optional)')
+    # Check for h5 suffix if '--file-prefix' argument used
+    if parser.parse_args().file_prefix is not None and \
```
Don't `parse_args()` multiple times. Insert this before line 90 and then use `args`:
```py
args = parser.parse_args()
```

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    if parser.parse_args().file_prefix is not None and \
+       parser.parse_args().file_prefix.endswith('.h5'):
+        print("Retry without .h5 extension on 'file-prefix'")
+        sys.exit(1)
```
Please replace lines 91-94 with an `assert`:
```py
assert args.file_prefix is None or args.file_prefix.endswith('.h5'), "Retry without .h5 extension on 'file-prefix'"
```

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+        print("Retry without .h5 extension on 'file-prefix'")
+        sys.exit(1)
+
+    return vars(parser.parse_args())
```
```py
return vars(args)
```

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+        h5_file_names = glob.glob(file_prefix + "*.h5")
+        if not h5_file_names:
+            sys.exit("Could not find files with prefix %s in path" %
+                     file_prefix)
```
```suggestion
assert len(h5_file_names) > 0, "Found no files with prefix {}.".format(file_prefix)
```

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+                h5_files.append(h5py.File(file_name, 'r'))
+            except OSError:
+                exit("Could not find file %s in directory %s" %
+                     (file_name, os.path.dirname(file_name)))
```
Instead of the `try except` you can `assert` if the file exists before trying to open it:

```py
assert os.path.exists(file_name), "File {} does not exist.".format(file_name)
```

wthrowe replied:
Don't do this.  You don't know all the reasons reading the file might fail, so you can't check them all beforehand.  (And I've seen programs get this wrong the other way, which is even more annoying: giving up when the file access would have succeeded.)  Even if it exists, the file could be a directory, or you might not have permission to read it.

nilsvu replied:
In those cases the exception will be printed and it gives you the reason, right?

wthrowe replied:
Sure, but that also happens if the file doesn't exist, doesn't it?  We should handle the errors consistently.

Backtraces are not a good way to tell the user they passed a bad command line argument.

nilsvu replied:
What's your suggestion then? The current `except` just hides the backtrace and always says "file not found" instead.

wthrowe replied:
I assume python has something similar to `std::exception::what()`.

nilsvu replied:
Yes, that's a good idea. Perhaps this is a good way to do it:

```py
try:
    h5_files.append(h5py.File(file_name, 'r'))
except OSError as err:
    sys.exit("Could not open file '{}': {}".format(file_name, err))
```

wthrowe replied:
That looks reasonable.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    return h5_files
+
+
+def print_var_names(args):
```
Take only the argument needed by this function:

```suggestion
def print_var_names(files):
```

Then pass it on to `get_h5_files`.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    :return list of h5 files containing volume data
+    '''
+    h5_files = []
+    if file_prefix is not None:
```
Take only one argument, e.g. name it `files`. Then decide what to do based on what that variables holds, e.g. like this:

```py
if isinstance(files, str):
    # Treat `files` as a prefix
else:
    # Treat `files` as a list of files
```

This is how such arguments are handled in other Python packages, e.g. Numpy

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    variables.remove("InertialCoordinates_x")
+    variables.remove("total_extents")
+    variables.remove("grid_names")
+    print("Variables in H5 file:\n[%s]" % ', '.join(map(str, variables)))
```
```suggestion
    print("Variables in H5 file:\n{}".format(list(map(str, variables))))
```

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    return None
+
+
+def get_data(args):
```
Make this function take only the arguments it needs, e.g. `def get_data(files, var_name)`

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+                 for obs_id in volfiles[0].keys()]
+    ids_times.sort(key=lambda pair: pair[1])
+    time_data = {}
+    for id_and_time in ids_times:
```
```suggestion
    for obs_id, local_time in ids_times:
```

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+        time_data[id_and_time].sort(key=lambda pair: pair[0])
+        time.append(id_and_time[1])
+        coords.append([pair[0] for pair in time_data[id_and_time]])
+        data.append([pair[1] for pair in time_data[id_and_time]])
```
you're appending to the `time_data` dictionary but never use it. I think you can just delete `time_data`. Also there's a lot of list manipulation going on, some of which is unnecessary. I suggest this:

```suggestion
        local_data = np.array(volfiles[0][obs_id][var_name])
        local_coords = np.array(volfiles[0][obs_id]['InertialCoordinates_x'])
        for volfile in volfiles[1:]:
            local_data = np.hstack([local_data, volfile[obs_id][var_name]])
            local_coords = np.hstack(
                [local_coords, volfile[obs_id]['InertialCoordinates_x']])
        ordering = np.argsort(local_coords)
        time.append(local_time)
        coords.append(local_coords[ordering])
        data.append(local_data[ordering])
```

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    return time, coords, data
+
+
+def render_single_time(args, coords, data):
```
again, just take the arguments you need: `render_single_time(time, coords, data, var_label, save_to=None)`

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    plt.plot(coords[int(args['time'])], data[int(args['time'])], 'o')
+    if args['save']:
+        print("Writing still to file %s.pdf" % args['save'])
+        plt.savefig("%s.pdf" % args['save'],
```
```suggestion
        plt.savefig(save_to + ".pdf",
```

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    if args['save']:
+        print("Writing still to file %s.pdf" % args['save'])
+        plt.savefig("%s.pdf" % args['save'],
+                    transparent=True,
```
Why do you make the background transparent? We can't include all possible styling possibilities in this script, so if a person wants to plot the data and then style it differently they can import and call these functions and add their styling code in their script. By default this should probably not be transparent, and neither use "tight layout".

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+        '''
+        Update the animation canvas
+        '''
+        title.set_text("t = %.5f" % time[iteration])
```
```suggestion
        title.set_text("t = {:.5f}".format(time[iteration]))
```

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    ax.set_xlabel('x')
+    ax.set_ylabel(args['var'])
+    title = ax.set_title("")
+    my_interval = float(args['interval'])
```
You're not using the `fps` from the command line args. Either get rid of the `interval` argument and use `fps` here
```suggestion
    interval_in_ms = 1000 / fps
```
or the other way round, i.e. get rid of the `fps` argument.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    if args['output']:
+        myfps = args['fps']
+        print("Writing animation to file %s.mp4 at %d frames per second" %
+              (args['output'], myfps))
```
also use `logging` here, see https://github.com/sxs-collaboration/spectre/pull/1578/files#r389366402

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    return None
+
+
+def render_animation(args, time, coords, data):
```
take only the arguments this function needs, see comment on the other functions

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+
+
+if __name__ == "__main__":
+    main(parse_cmd_line())
```
You don't need the extra function `main`. You can move the body of that function here.

nilsvu reviewed (commented):
Nice, I can run an elliptic solve and plot it right away with this script 👍 This is very useful.

<img width="752" alt="Bildschirmfoto 2020-03-26 um 15 40 59" src="https://user-images.githubusercontent.com/746230/77660563-c7152880-6f79-11ea-88f5-600712f22823.png">

nilsdeppe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,269 @@
+                                   frames=len(time),
+                                   interval=my_interval)
+    if output_prefix:
+        my_fps = 1000/my_interval
```
you probably want this to be `1000.0` to make sure things happens as floating point division

nilsdeppe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,269 @@
+        sys.exit(0)
+    time, coords, data = get_data(files, args['var'])
+    if args['interval'] is None:
+        interval = 1000/args['fps']
```
you probably want this to be `1000.0` to make sure things happens as floating point division

kidder commented:
@isha1810 please format the file using `yapf` which can be run with ` yapf -i PYTHON_FILE.py`

kidder commented:
@isha1810 you can squash the fixup

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+try:
+    import Tkinter as tkinter
+except ImportError:
+    import tkinter
```
Why is the `try/except` necessary? Please add a comment

isha1810 replied:
The module name depends on the python version. I have added a comment above.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+                             type=str,
+                             nargs='+',
+                             help='List of all files if they do not '
+                             'have same file prefix. You must include \'.h5\' '
```
Just an FYI (not a change request): you can use different quotation mark styles to avoid the escaping with `\`. E.g. if you use `"` for the string then you can use `'` within it.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+                            help='Delay between frames in  milliseconds')
+    # Check for h5 suffix if '--file-prefix' argument used
+    args = parser.parse_args()
+    assert args.file_prefix is None or not args.file_prefix.endswith('.h5'),\
```
Another FYI (not a change request): It can be useful to wrap code such as the condition of this assert or the string in parentheses `()`. Then they can span multiple lines, you don't need the `\` and `yapf` will format it automatically.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+    print("Variables in H5 file:\n{}".format(list(map(str, variables))))
+    for h5_file in h5files:
+        h5_file.close()
+    return None
```
delete this line

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+        ordering = None
+        for volfile in volfiles:
+            try:
+                local_data = local_data +\
```
FYI, the trick with the parentheses to break long lines works also for operator expressions like this

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+        plt.savefig(output_prefix + ".pdf", format='pdf')
+    else:
+        plt.show()
+    return None
```
delete this line

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+        Initialize the animation canvas
+        :return: empty line and title info
+        '''
+        line.set_data([], [])
```
before this line add:
```py
global line, title
```
(same in the `animate` function below)

isha1810 replied:
When I add these lines I get the following error:
`NameError: global name 'line' is not defined `

nilsvu replied:
you should move the commands where you initialize `line` and `title` up, so they run before the `init` and `animate` functions are defined. I add a comment to the relevant code change below.

isha1810 replied:
This still gives me the `NameError` probably because `line`, `title` are initialized in the `render_animation` function and not globally. I could also try `nonlocal` but this is different in Python 2 and 3. Is there anything else I can try?

nilsvu replied:
You're right that `nonlocal` would be the correct choice, but it doesn't exist in Python 2. Let's leave it as-is then, without a `global` or `nonlocal` statement. As long as you don't assign to the `line` or `title` variables but only call member functions on them the code should be fine. You can delete lines 206 and 207.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+    return None
+
+
+def render_animation(var_name, output_prefix, my_interval, time, coords, data):
```
delete the `my_` prefix

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+                                   frames=len(time),
+                                   interval=my_interval)
+    if output_prefix:
+        my_fps = 1000 / my_interval
```
delete `my_` prefix

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,279 @@
+        my_fps = 1000 / my_interval
+        logging.info(
+            "Writing animation to file %s.mp4 at %d frames per second" %
+            (output_prefix, my_fps))
```
Prefer `.format()` (%-formatting is old and annoying)

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,277 @@
+    import matplotlib.pyplot as plt
+    import matplotlib.animation as animation
+    line = ()
+    title = None
```
Delete these two new lines 206--207. Instead, move the lines 209--224 down, above line 234. Then add the `global line, title` statement to both functions.

nilsvu reviewed (commented):
Please remember to keep fixup commits around until the reviewers have had a chance to look at them. Else it's more difficult for the reviewers to understand what part of the code changed :)

kidder commented:
@isha1810 @nilsdeppe @nilsleiffischer what needs to be done next for this?

isha1810 commented:
> @isha1810 @nilsdeppe @nilsleiffischer what needs to be done next for this?

I think I've addressed all the comments. 
@nilsdeppe @nilsleiffischer what do you think?

kidder commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,274 @@
+def print_var_names(files):
+    '''
+    Print all available variables to screen
+    :param args:
```
what does this do?  should it list files?

nilsvu replied:
@kidder I think the function and docstring are reasonably descriptive. But @isha1810 you changed the argument to the function from `args` to just the args it needs, i.e. `files`, but didn't update the documentation. Please change `:param args:` to `:param files:` and describe the expected argument `files`, e.g. "List of file names or common file name prefix".

(the same applies to Larry's other 3 comments)

kidder commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,274 @@
+def render_single_time(var_name, time_slice, output_prefix, coords, data):
+    '''
+    Renders a single time snap shot
+    :param args: the command line arguments
```
I don't know what this does

kidder commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,274 @@
+def render_animation(var_name, output_prefix, interval, time, coords, data):
+    '''
+    Render an animation of the data
+    :param args: the command line arguments
```
what does this do?

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,274 @@
+def get_h5_files(files):
+    '''
+    Get a list of the h5 files containing the data
+    :return list of h5 files containing volume data
```
`:return:` (you're missing a `:`)

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,274 @@
+def get_data(files, var_name):
+    '''
+    Get the data to be plotted
+    :return: the array of data, coords, and time
```
The order of return values you describe here does not match the order they are actually returned in, please correct.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,274 @@
+    for obs_id, local_time in ids_times:
+        local_coords = []
+        local_data = []
+        ordering = None
```
delete this line

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,274 @@
+                            "or 'Error(psi)'")
+    group_vars.add_argument(
+        '--list-vars',
+        type=bool,
```
change this line to `action='store_true'`. Then no argument needs to be given in the command line, but specifying `--list-vars` means "true".

Even better would be if you just removed the `--list-vars` argument and have the script print the available variables if the `--var` argument is not given. I don't request you change that though, do what you prefer.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,274 @@
+
+    plt.xlabel("x")
+    plt.ylabel(var_name)
+    plt.title("t = {:.5f}".format(int(time_slice)))
```
Don't convert to `int`. This line currently always displays 5 zeros after the decimal point. Just remove the `int()` so the 5 numbers after the decimal point are actually meaningful, or change the `.5f` to the precision you want.

nilsvu commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,274 @@
+    if args['interval'] is None:
+        interval = 1000 / args['fps']
+    if args['time'] is None:
+        render_animation(args['var'], args['output'], interval, time, coords,
```
This line gives an error if the script is run with `--interval`. Add an `else` clause to the `if args['interval'] is None` where you do this:
```py
interval = args['interval']
```

nilsvu reviewed (changes requested):
Other than the few minor comments below, the one thing this PR is missing is a test. You don't have to test everything, just that the script runs for a few example parameters. Look at `tests/Unit/Visualization/Python/Test_GenerateXdmf.py` to see how to write such a test.

nilsvu reviewed (commented):
The fixup looks good to me, please squash it. From my perspective, only the test is now missing. I don't think you necessarily have to wait for #2309 and #2318 to be able to write tensor data, but a small test like GenerateXdmf has would suffice for now.

kidder commented on tests/Unit/Visualization/Python/Test_Render1D.py:
```diff
@@ -0,0 +1,50 @@
+        self.assertTrue(os.path.isfile(output_prefix + '.pdf'))
+        os.remove(output_prefix + '.pdf')
+
+    def test_render_animation(self):
```
remove this and squash into single commit

nilsvu replied:
Do you mean remove line 22?

nilsdeppe replied:
no, remove `test_render_animation`

nilsvu commented on tests/Unit/Visualization/Python/Test_Render1D.py:
```diff
@@ -0,0 +1,50 @@
+        self.assertEqual(find_extrema_over_data_set(test_array), expected_vals)
+
+
+#    def test_parse_cmd_line(self):
```
delete this line

nilsvu reviewed (dismissed):
Looks good to me. You can squash the commit that adds the test into the other one.

nilsdeppe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,289 @@
+# The module name below depends on python version used to run the script
+# Imports Tkinter for python2, tkinter for python3
+try:
+    import Tkinter as tkinter
```
I don't know how to install tkinter in the container with pip. Can you either figure that out or not use tkinter (I don't actually know what it is or does, so maybe you have to use it)?

nilsvu replied:
I think it can just be removed, I never needed tkinter to render animations.
It's used on line 261 below, but that whole else clause can probably go and perhaps be replaced by an assert somewhere.

isha1810 replied:
I can remove it.

nilsdeppe reviewed (commented):
Looks good! Please go ahead and squash this and rebase on develop :)

nilsdeppe reviewed (dismissed):
@kidder @nilsleiffischer @wthrowe could you please take a look at this to make sure your comments have been addressed? Thanks!

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,275 @@
+import os
+import numpy as np
+import logging
+import matplotlib as mpl
```
matplotlib is a new dependency for SpECTRE and should go in the installation instructions.

wthrowe replied:
ping

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,275 @@
+    '''
+
+    h5files = get_h5_files(files)
+    volfile = h5files[0]["element_data.vol"]
```
I don't think "element_data.vol" is a standard name anymore, so this should be specifiable.  The PlaneWave1DObserveExample.yaml example produces "VolumePsi0And100.vol" and "VolumePsiPiPhiEvery50Slabs.vol".

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,275 @@
+
+    plt.xlabel("x")
+    plt.ylabel(var_name)
+    plt.title("t = {:.5f}".format(time[time_slice]))
```
This throws an IndexError if time_slice is too large, and I don't think there is any way for the user to learn the valid range.

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -0,0 +1,275 @@
+                            list(volfile[obs_id]['InertialCoordinates_x']))
+        ordering = np.argsort(local_coords)
+        coords.append(np.array(local_coords)[ordering])
+        data.append(np.array(local_data)[ordering])
```
--list-vars suggested "bases" as a valid variable (that's probably another bug), but when I requested it I got the unclear error on this line: "IndexError: index 4 is out of bounds for axis 0 with size 4"

wthrowe commented on tests/Unit/Visualization/Python/CMakeLists.txt:
```diff
@@ -5,3 +5,8 @@ spectre_add_python_bindings_test(
   Test_GenerateXdmf.py
   "unit;visualization;python")
+
+spectre_add_python_bindings_test(
```
I don't really know how the python tests work, but this isn't a bindings test, is it?

wthrowe reviewed (changes requested):
This looks much better, but I'd still like to see some improvements.

nilsdeppe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -44,6 +44,10 @@ def parse_cmd_line():
                              help="List of all files if they do not "
                              "have same file prefix. You must include '.h5' "
                              "extension ")
+    parser.add_argument('--vol-file', type=str, required=True,
```
could you rename this argument to `--subfile-name` to make `GenerateXdmf`? It'll be easier on folks if options are named consistently across files :)

nilsdeppe reviewed (commented):
One minor naming suggestion for consistency.

@wthrowe could you please take a look at the fixups when you get a chance?

I'm happy for you to squash once @wthrowe has had a chance to look at the fixups, including my minor rename suggestion :)

wthrowe commented on src/Visualization/Python/Render1D.py:
```diff
@@ -44,6 +44,10 @@ def parse_cmd_line():
                              "have same file prefix. You must include '.h5' "
                              "extension ")
+    parser.add_argument('--vol-file', type=str, required=True,
+                        help="Name of subfile within h5 file containing"
```
Missing space at end of string.

wthrowe reviewed (commented):
You can squash.

isha1810 committed:
add matplotlib to docs

nilsdeppe commented:
looks good to me. @wthrowe could you take a look at the fixup and the new commit when you get a chance?

wthrowe commented:
Looks fine.

nilsdeppe commented:
@isha1810 do you understand the test failure?

isha1810 commented:
@nilsdeppe @wthrowe I added Test_GenerateXdmf using `spectre_add_python_bindings_test` in the fixup, instead of `spectre_add_python_test` (which was causing some checks to fail because test_GenerateXdmf depends on Informer bindings).

nilsdeppe commented:
Looks good! #2631 discusses the issue and presents a few solutions on how to fix it.

nilsdeppe commented:
@wthrowe @kidder could you please look at this?

wthrowe reviewed (approved):
I still think throwing uncaught exceptions for user errors is horrible UI, but if that's the python way of doing things so be it.

nilsdeppe commented:
@wthrowe I agree... Biggest reason I dislike python and am not convinced it's actually easier to learn than C++... At least compilers give (mostly) reasonable errors, whereas python you just get a bloody stack trace...
